middleware/file/lookup.go

```diff
@@ -95,6 +95,22 @@ func (z *Zone) Lookup(state request.Request, qname string) ([]dns.RR, []dns.RR,
 			continue
 		}
 
+		// If we see DNAME records, we should return those.
+		if dnamerrs := elem.Types(dns.TypeDNAME); dnamerrs != nil {
+			// There can only be one?
+			dname := dnamerrs[0]
+			if cname := synthesizeCNAME(state.Name(), dname.(*dns.DNAME)); cname != nil {
+				answer, ns, extra, rcode := z.searchCNAME(state, elem, []dns.RR{cname})
+
+				// The relevant DNAME RR should be included in the answer section,
+				// if the DNAME is being employed as a substitution instruction.
+				answer = append([]dns.RR{dname}, answer...)
+
+				return answer, ns, extra, rcode
+			}
+			// What should happen here? Just the DNAME w/ SIG?
```

Comment:
The domain name that owns a DNAME record is allowed to have other RR types at that domain name, except those have restrictions on what they can coexist with (e.g. another DNAME). So there is nothing special left here.